packages/tenant-process/src/services/tenantService.ts

```diff
@@ -369,4 +461,71 @@ export function tenantServiceBuilder(
     },
   };
 }
+
+async function assignCertifiedAttribute({
+  targetTenant,
+  attribute,
+  readModelService,
+}: {
+  targetTenant: Tenant;
+  attribute: Attribute;
+  readModelService: ReadModelService;
+}): Promise<Tenant> {
+  const certifiedTenantAttribute = targetTenant.attributes.find(
+    (attr): attr is CertifiedTenantAttribute =>
+      attr.type === tenantAttributeType.CERTIFIED && attr.id === attribute.id
+  );
+
+  // eslint-disable-next-line functional/no-let
+  let updatedTenant: Tenant = {
+    ...targetTenant,
+    updatedAt: new Date(),
+  };
+
+  if (!certifiedTenantAttribute) {
+    // assigning attribute for the first time
+    updatedTenant = {
+      ...updatedTenant,
+      attributes: [
+        ...targetTenant.attributes,
+        {
+          id: attribute.id,
+          type: tenantAttributeType.CERTIFIED,
+          assignmentTimestamp: new Date(),
+          revocationTimestamp: undefined,
+        },
+      ],
+    };
+  } else if (!certifiedTenantAttribute.revocationTimestamp) {
+    throw certifiedAttributeAlreadyAssigned(attribute.id, targetTenant.id);
+  } else {
+    // re-assigning attribute if it was revoked
+    updatedTenant = {
+      ...updatedTenant,
+      attributes: targetTenant.attributes.map((a) =>
+        a.id === attribute.id
+          ? {
+              ...a,
+              assignmentTimestamp: new Date(),
+              revocationTimestamp: undefined,
+            }
+          : a
+      ),
+    };
+  }
+  const tenantKind = await getTenantKindLoadingCertifiedAttributes(
+    readModelService,
+    updatedTenant.attributes,
+    updatedTenant.externalId
+  );
+
+  if (updatedTenant.kind !== tenantKind) {
+    updatedTenant = {
+      ...updatedTenant,
+      kind: tenantKind,
+    };
+  }
```

Comment:
no strong opinions, but the `kind` change is falling inside the `CreateEventTenantCertifiedAttributeAssigned`, which makes it hard to intercept.

@beetlecrunch thoughts?

Reply:
@galales, this change could be very useful (for example: sending a specific notification), so it would be appropriate to track it with an ad-hoc event.
This means that sometimes we should handle two different events in the same transaction.
The alternative is to use a single event to test for variation, but we need to bring that variation into the event itself.

Reply:
ok thanks! let's go for the double event, so that it could be re-used in case of future business logics